modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/reducer.ts

```diff
@@ -22,7 +22,19 @@ const STATES = {
 
 export {STATES};
 
-export default function reducer(state, action) {
+interface State {
+	current: {show: boolean};
+	nextTarget?: HTMLElement;
+	target?: HTMLElement;
+	timestamp?: number;
```

Comment:
I'm not quite sure the best to to type this. For the argument we pass to the `reducer(` function, the arguments are optional but when we return state from the reducer it will always exist. Is there a more appropriate way to do this? My ts brain is a bit foggy from parental leave

Reply:
If I understand your question correctly, you're saying how we initialize with `{current: STATES.IDLE}` but always set the other properties in our `reducer()` return value, right?

"Always" isn't strictly true, because there is one case we we reset back and:

```ts
return {current: STATES.IDLE};
```

and a couple others where we could conceivably "trick" the reducer into returning an incomplete set of properties, if we dispatched actions in the right order:

```ts
return {
	...state,
	timestamp: Date.now(),
};
```

For me when this kind of situation arises the question isn't "How do I best express this in the type system?" as "Is the pain of typing this a sign that I should be refactoring instead?". In this case, I think it might be worth at least considering that (making sure all the properties are at least present, even if the values are undefined).

Having said that though, I don't actually think what you have here looks too bad, and I really love the way you were able to port 110-line `reducer.js` and 229-line `index.js` to TS with _barely any explicit type annotations at all_. That's awesome. (No doubt some of this is because of the `any` types from `frontend-js-web`, but still...)